lib/salus/scanners/report_node_modules.rb

```diff
@@ -58,7 +62,72 @@ def self.supported_languages
 
     private
 
+    def yarn_version
+      shell_return = run_shell(YARN_VERSION_COMMAND)
+      # stdout looks like "1.22.0\n"
+      shell_return.stdout&.strip
+    end
+
     def record_dependencies_from_yarn_lock
+      if Gem::Version.new(yarn_version) >= Gem::Version.new(BREAKING_VERSION)
+        record_dependencies_from_latest_yarn_lock
+      else
+        record_dependencies_from_legacy_yarn_lock
+      end
+    end
+
+    def record_dependencies_from_latest_yarn_lock
+      packages_shell_return = run_shell(
+        YARN_LIST_PACKAGES_COMMAND, chdir: @repository.path_to_repo
+      )
+      unless packages_shell_return.success?
+        raise ParseError, "Failed to parse yarn.lock file: #{packages_shell_return.stderr}"
+      end
+
+      source_shell_return = run_shell(
+        YARN_REGISTRY_SOURCE_COMMAND, chdir: @repository.path_to_repo
+      )
+      unless source_shell_return.success?
+        raise ParseError, "Failed to fetch npmRegistryServer: #{source_shell_return.stderr}"
+      end
+
+      dependencies = []
+      # source shell output - https://registry.yarnpkg.com
+      source = source_shell_return.stdout.strip
+
+      yarn_packages = packages_shell_return.stdout
+      # yarn info shell output -
+      # "gulp-cssmin@npm:0.2.0"
+      # "gulp-modify-file@npm:1.0.1"
+      yarn_packages.each_line do |line|
+        name = line.split("@npm")[0]
+        version = line.split("@npm")[1]
+        if name.present? && version.present?
+          # Handle patched packages. Sample format:
+          # "node-fetch-web-streams@patch:node-fetch-web-streams@npm%3A1.0.1#.yarn/patches/
+          # node-fetch-web-streams-npm-1.0.1-63d9cb5d48.patch::version=1.0.1&hash=804812&locator=%40
+          # frontend%2Fmobile%40workspace%3A."
+          version = version[/\::version=(.*?)&/, 1] if version.include?("::version=")
```

Comment:
What is `(.*?)` supposed to mean?  
It seems to mean 0 or 1 of "0 or more of any character", which seems a little confusing to me.

Reply:
Mainly to match non greedy, since I only want the first match.